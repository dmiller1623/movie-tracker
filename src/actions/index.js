export const populateMovies = (movies) => ({
  type: 'POPULATE_MOVIES',
  movies
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
});

export const signUpUser = (email, password) => ({
  type: 'SIGNUP_USER',
  email,
  password
});

export const signOutUser = () => {
  type: 'SIGN_OUT'

}
