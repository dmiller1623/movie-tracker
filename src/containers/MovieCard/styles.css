article {
  width: 285px;
  margin: .5rem; }
  article img {
    width: 285px;
    border-radius: 10px; }

svg {
  position: absolute;
  margin-left: calc(285px - 4.2rem);
  transition: all .3s;
  cursor: pointer; }
  svg:hover {
    width: 4.2rem; }

.info {
  position: absolute;
  padding-top: 3.5rem;
  width: 285px;
  border-radius: 10px;
  color: #f2f2f2;
  background-color: #1072B9;
  text-align: left;
  opacity: 1;
  transition: opacity .25s ease-in-out;
  -moz-transition: opacity .25s ease-in-out;
  -webkit-transition: opacity .25s ease-in-out; }
  .info p {
    margin: .5rem; }
  .info .overview {
    font-family: 'Ubuntu Condensed', sans-serif; }

.hide {
  opacity: 0;
  visibility: hidden; }

@media only screen and (max-width: 430px) {
  article {
    width: 250px; }
    article img {
      width: 250px; }
    article svg {
      margin-left: calc(250px - 4.2rem);
      transition: none; }
  .info {
    width: 250px; } }
