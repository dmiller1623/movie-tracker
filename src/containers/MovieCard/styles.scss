article {
  width: 285px;
  margin: .5rem;
  img {
    width: 285px;
    border-radius: 10px;
  }
}

svg {
  position: absolute;
  margin-left: calc(285px - 4.2rem);
  transition: all .3s;
  cursor: pointer;
  &:hover {
    width: 4.2rem;
  }
}

@media only screen and (max-width: 430px) { 
  article {
    width: 250px;
    img {
      width: 250px;
    }
    svg {
      margin-left: calc(250px - 4.2rem);
      transition: none;
    }
  } 
}