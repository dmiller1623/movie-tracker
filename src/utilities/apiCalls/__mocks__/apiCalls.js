export const getMovies = jest.fn().mockImplementation(() => ({}));

export const getUser = jest.fn().mockImplementation(() => ({}));

export const getFavorites = jest.fn().mockImplementation(() => ({}));

export const addNewUser = jest.fn().mockImplementation(() => ({}));

export const addNewFavorite = jest.fn().mockImplementation(() => ({}));

export const deleteFavorite = jest.fn().mockImplementation(() => ({}));