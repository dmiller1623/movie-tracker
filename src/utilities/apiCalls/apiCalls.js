import apiKey from '../../apiKey.js';
import cleanData from '../helper/helper.js';

export const getMovies = async () => {
  const url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&with_genres=878&language=en-US&sort_by=revenue.asc&include_adult=false&include_video=false&page=1`;
  try {
    const response = await fetch(url);
    const movieData = await response.json() ;
    const unresolvedMovies = await cleanData(movieData);
    return Promise.all(unresolvedMovies);
  } catch (error) {
    throw new Error(error.message);
  }
};

export const getUser = async (email, password) => {
  try {
    const response = await fetch('http://localhost:3000/api/users', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      email,
      password
    })
  })
  if (response.status !== 200) {
    alert('User does not exist, please try again or sign up for an account')
    return {};
  } else {
    const results = await response.json();
    return results.data
  }
  } catch(error) {
    throw new Error(error.message)
  }
  
};

export const addNewUser = async (name, email, password) => {
  try {
    const response = await fetch('http://localhost:3000/api/users/new', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        name, 
        email,
        password
      })
    })
    if (response.status !== 200) {
      alert('User already exists, please try another email or login to an account')
      return {};
    } else {
      const results = await response.json();
      return results.data
    }  
  } catch(error) {
    throw new Error(error.message)
  }
};

