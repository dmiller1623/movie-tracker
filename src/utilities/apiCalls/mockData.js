export const mockMovieDataFromFetch = {
  "page": 1,
  "total_results": 9325,
  "total_pages": 467,
  "results": [
      {
          "vote_count": 3,
          "id": 72803,
          "video": false,
          "vote_average": 10,
          "title": "SPFX: The Empire Strikes Back",
          "popularity": 0.384,
          "poster_path": "/f0sS931i5xq7Xo0YS035OfdYoLN.jpg",
          "original_language": "en",
          "original_title": "SPFX: The Empire Strikes Back",
          "genre_ids": [
              99,
              878
          ],
          "backdrop_path": null,
          "adult": false,
          "overview": "A behind-the-scenes look at the creation of the special effects in the second Star Wars installment.",
          "release_date": "1980-09-22"
      },
      {
          "vote_count": 1,
          "id": 77764,
          "video": false,
          "vote_average": 10,
          "title": "Playmobil: Future Planet",
          "popularity": 0.22,
          "poster_path": "/blKu0AnRxXIMAv17OLnx7fIvS2g.jpg",
          "original_language": "de",
          "original_title": "Playmobil: Future Planet",
          "genre_ids": [
              878,
              10751,
              16
          ],
          "backdrop_path": "/hMXD9GID3QeI9PheB8BS6bEiALp.jpg",
          "adult": false,
          "overview": "",
          "release_date": "2011-01-01"
      },
      {
          "vote_count": 1,
          "id": 65146,
          "video": false,
          "vote_average": 10,
          "title": "Big Meat Eater",
          "popularity": 0.092,
          "poster_path": "/gimNITXaKUevTsbXNUFEI6bCqro.jpg",
          "original_language": "en",
          "original_title": "Big Meat Eater",
          "genre_ids": [
              35,
              878
          ],
          "backdrop_path": "/mNJW7JV55hOqyfzfA07132BLgRH.jpg",
          "adult": false,
          "overview": "After killing the crooked mayor a homicidal janitor named Abdullah goes to work for a butcher who has invented a new language for the town's planned futuristic theme park. In the butcher shop is a septic tank where scraps of meat are dumped which has produced \"balonium\" a radioactive fuel source sought after by space aliens. The aliens revive the mayor, who's body is hanging in the meat locker of the butcher shop, in an attempt to gain access to the balonium.",
          "release_date": "1982-04-30"
      }
    ]
};

export const movieDataResult = [
    {
        movie_id: 0,
        title: "SPFX: The Empire Strikes Back",
        poster_path: "/f0sS931i5xq7Xo0YS035OfdYoLN.jpg",
        release_date:  "1980-09-22",
        vote_average: 10,
        overview: "A behind-the-scenes look at the creation of the special effects in the second Star Wars installment."  
    },
    {
        movie_id: 1,
        title: "Playmobil: Future Planet",
        poster_path:  "/blKu0AnRxXIMAv17OLnx7fIvS2g.jpg",
        release_date: "2011-01-01",
        vote_average: 10,
        overview: "" 
    },
    {
        movie_id: 2,
        title: "Big Meat Eater",
        poster_path: "/gimNITXaKUevTsbXNUFEI6bCqro.jpg",
        release_date: "1982-04-30",
        vote_average: 10,
        overview: "After killing the crooked mayor a homicidal janitor named Abdullah goes to work for a butcher who has invented a new language for the town's planned futuristic theme park. In the butcher shop is a septic tank where scraps of meat are dumped which has produced \"balonium\" a radioactive fuel source sought after by space aliens. The aliens revive the mayor, who's body is hanging in the meat locker of the butcher shop, in an attempt to gain access to the balonium." 
    }
];