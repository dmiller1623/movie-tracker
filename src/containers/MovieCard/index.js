import React, { Component } from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';

import { addFavorite, populateFavorites } from '../../actions';
import { addNewFavorite, deleteFavorite } from '../../utilities/apiCalls/apiCalls';

import './styles.css';

export class MovieCard extends Component {
  toggleFavorite = async () => {
    const { id, name } = this.props.user;
    const { favorites, addFavorite, populateFavorites } = this.props;
    const movie = {...this.props, user_id: id};
    
    if (!name) {
      return alert('To favorite a movie, please login or sign up for a new account.');
    }

    if (favorites.includes(movie.movie_id)) {
      await deleteFavorite(id, movie.movie_id);
      const updatedFavorites = favorites.filter( favorite => favorite !== movie.movie_id);
      populateFavorites(updatedFavorites);
    } else {
      await addNewFavorite(movie);
      addFavorite(movie.movie_id);
    }
  }

  render() {
    const { title, poster_path, starColor} = this.props;
    const posterUrl = `http://image.tmdb.org/t/p/w342//${poster_path}`;

    return (
      <article>
        <svg 
          onClick={() => this.toggleFavorite()}
          viewBox='0 0 512 512' width='2.5rem'>
          <path d='m463.515 0h-415.03v512l207.515-130.554 207.515 130.554z' fill='#f14742'/>
          <path d='m388.692 162.487c-.882-2.716-3.758-5.228-8.513-5.515l-84.734-5.664-31.571-78.837c-1.742-4.433-5.021-6.392-7.876-6.392s-6.134 1.959-7.876 6.392l-31.571 78.837-84.734 5.664c-4.755.287-7.631 2.799-8.513 5.515-.882 2.715-.032 6.439 3.645 9.465l65.222 54.387-20.797 82.336c-1.197 4.61.305 8.124 2.613 9.801 2.309 1.677 6.114 2.018 10.129-.544l71.884-45.22 71.88 45.222c4.015 2.562 7.819 2.221 10.129.544 2.31-1.678 3.81-5.192 2.613-9.801l-20.797-82.336 65.221-54.387c3.679-3.028 4.529-6.753 3.646-9.467z' fill={starColor}/>
        </svg>
        <img src={posterUrl} alt={title + ' poster_path'} />
      </article>
    );
  }
}

export const mapStateToProps = (state) => ({
  user: state.user,
  favorites: state.favorites
});

export const mapDispatchToProps = (dispatch) => ({
  addFavorite: (movieId) => dispatch(addFavorite(movieId)),
  populateFavorites: (favorites) => dispatch(populateFavorites(favorites))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);

const { string, func, number, shape, arrayOf } = PropTypes;
MovieCard.propTypes = {
  title: string,
  poster_path: string,
  starColor: string,
  favorites: arrayOf(number),
  user: shape({
    id: number,
    name: string
  }),
  addFavorite: func,
  populateFavorites: func
};