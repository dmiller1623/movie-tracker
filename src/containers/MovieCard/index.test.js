import React from 'react';
import { shallow } from 'enzyme';
import {  MovieCard, mapStateToProps, mapDispatchToProps } from '../MovieCard';
import { addFavorite, populateFavorites } from '../../actions';
import { addNewFavorite, deleteFavorite } from '../../utilities/apiCalls/apiCalls';


jest.mock('../../utilities/apiCalls/apiCalls.js');

describe('MovieCard', () => {
  let wrapper;

  describe('MovieCard component', () => {
    const movie = {movie_id: 3, title: "Skulhedface", poster_path: "/1hOcS0G9q5eEzJEYeWjAa376yux.jpg", release_date: "1994-01-01", vote_average: 10};
    const populateFavorites = jest.fn();
    const addFavorite = jest.fn();
    const starColor = '#ffd24d';

    it('should render a movieCard with the correct starColor', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should return an alert if toggleFavorite is called and user is not logged in', async () => {
      const user = {};
      const favorites = [0, 1, 2, 3];
      const mockEvent = {stopPropagation: jest.fn()};

      wrapper = shallow(<MovieCard {...movie} starColor={starColor} user={user} favorites={favorites} addFavorite={addFavorite} populateFavorites={populateFavorites}/> );

      wrapper.instance().toggleFavorite(mockEvent);

      expect(deleteFavorite).not.toHaveBeenCalled();

    });

    it('should call deleteFavorite and populateFavorite with the correct params if toggleFavorite is called on a movie that is already in the users favorites', () => {
      const user = {id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com"};
      const favorites = [0, 1, 2, 3];
      const mockEvent = {stopPropagation: jest.fn()};

      wrapper = shallow(<MovieCard {...movie} starColor={starColor} user={user} favorites={favorites} addFavorite={addFavorite} populateFavorites={populateFavorites}/> );

      wrapper.instance().toggleFavorite(mockEvent);

      expect(deleteFavorite).toHaveBeenCalledWith(1, 3);
    });

    it('should call addNewFavorite and addFavorite with the correct params if toggleFavorite is called on a movie that is not already in the user favorites', () => {
      const user = {id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com"};
      const favorites = [0, 1, 2];
      const expected = {"addFavorite": addFavorite, "favorites": [0, 1, 2], "movie_id": 3, "populateFavorites": populateFavorites, "poster_path": "/1hOcS0G9q5eEzJEYeWjAa376yux.jpg", "release_date": "1994-01-01", "starColor": "#ffd24d", "title": "Skulhedface", "user": {"email": "tman2272@aol.com", "id": 1, "name": "Taylor", "password": "password"}, "user_id": 1, "vote_average": 10};
      const mockEvent = {stopPropagation: jest.fn()};

      wrapper = shallow(<MovieCard {...movie} starColor={starColor} user={user} favorites={favorites} addFavorite={addFavorite} populateFavorites={populateFavorites}/> );
      wrapper.instance().toggleFavorite(mockEvent);

      expect(addNewFavorite).toHaveBeenCalledWith(expected);
    });

    it('should call toggleFavorite when svg is clicked', () => {
      const user = {id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com"};
      const favorites = [0, 1, 2];
      const mockEvent = {stopPropagation: jest.fn()};
      wrapper = shallow(<MovieCard {...movie} starColor={starColor} user={user} favorites={favorites} addFavorite={addFavorite} populateFavorites={populateFavorites}/> );

      const spy = jest.spyOn(wrapper.instance(), 'toggleFavorite');
      wrapper.find('svg').simulate('click', mockEvent);

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with a todos array', () => {
      const mockUser = {id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com"};
      const mockFavorites = [0, 1, 2];

      const mockState = {user: {id: 1, name: "Taylor", password: "password", email: "tman2272@aol.com"}, favorites: [0, 1, 2]};

      const expected = { user: mockUser, favorites: mockFavorites};

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using a function addFavorite', () => {
      const mockMovieId = '5';
      const mockDispatch = jest.fn();
      const actionToDispatch = addFavorite(mockMovieId);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addFavorite(mockMovieId);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when using a function populateFavorites', () => {
      const mockMovieId = '5';
      const mockDispatch = jest.fn();
      const actionToDispatch = populateFavorites(mockMovieId);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.populateFavorites(mockMovieId);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

    });
  });

});