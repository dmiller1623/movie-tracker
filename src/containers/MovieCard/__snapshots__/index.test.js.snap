// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard MovieCard component should render a movieCard with the correct starColor 1`] = `undefined`;
